📝 Enumerating eslint rules + package refactoring

camilleriluke opened the pull request:
This is a non breaking change.

Followed the basic structure of [eslint-config-walmart](https://github.com/walmartlabs/eslint-config-walmart) and migrated our rules.

**What changed**
* Split the rules into categories according to http://eslint.org/docs/rules/
* Added supports for different configurations (es5, es5-test, es6, etc.)
* Added support for `commitizen`
* Added validation on commit messages (`husky` and `validate-commit-msg`)
* Added `eslint` as a dependency and peer dependency
* Added npm scripts
  * `npm run commit` for those who don't have` commitizen` installed globally
  * `npm run lint` self linting
  * `npm run find-new-eslint-rules` - checks for new (missing) rules inside for all the configurations
  * `npm run test` - lints and checks for rules
* Pimped up `README`
* Added a license files
* Added ignore on debug files

## More things to look into
- [x] Add contribution guide
  - [x] How to commit
- [x] Split envs into separate files
- [x] Add support for `eslint-find-rules`
- ~Add travis support (maybe this shouldn't be part of this PR)~ Moved to issue https://github.com/Casumo/eslint-config-casumo/issues/2
- ~Add version bumping (maybe this shouldn't be part of this PR)~ Moved to issue https://github.com/Casumo/eslint-config-casumo/issues/3

Commit:
docs(license): Add license file

Commit:
refactor(.gitignore): Add more ignores

Ignoring debug files

Commit:
refactor(package.json): Update package.json tag values

Commit:
refactor(package.json): Add new dev dependencies

Added support for commitizen, and validation for commit messages. Also added eslint as a dev

dependency and peer dependency

Commit:
docs(README): Update readme

Updated readme to be more expressive

Commit:
refactor(rules): Enumerated all eslint rules

Added all rules according to http://eslint.org/docs/rules/

Commit:
docs(README): Update docs

Commit:
feat(rules): Add es6 eslint rules

leventebalogh commented on README.md:
```diff
@@ -1,16 +1,30 @@
+[![version](https://img.shields.io/npm/v/eslint-config-casumo.svg?style=flat-square)](https://www.npmjs.com/package/eslint-config-casumo)
+[![downloads](https://img.shields.io/npm/dm/eslint-config-casumo.svg?style=flat-square)](http://npm-stat.com/charts.html?package=eslint-config-casumo&from=2015-08-01)
+[![MIT License](https://img.shields.io/npm/l/eslint-config-casumo.svg?style=flat-square)](http://opensource.org/licenses/MIT)
+[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)
```
Nice, just wanted to recommend the commitizen-friendly badge, but see it's already there 😛

leventebalogh commented on best-practices.js:
```diff
@@ -0,0 +1,77 @@
+module.exports = {
```
I think it would be nice to move these rule categories to a sub-directory, like `/rules`

leventebalogh commented on best-practices.js:
```diff
@@ -0,0 +1,77 @@
+        browser: true,
+        node: true
+    },
+    rules: {
```
I think we shouldn't include any environment setup in the rules themselves to make them more composable.
So like:
```javascript
module.exports = {
    rules: {
        /* ... */
    }
}

Commit:
refactor(rules): Bundle es6 extenders for easy importing

leventebalogh commented on best-practices.js:
```diff
@@ -0,0 +1,77 @@
+        // NO_RULE 'require-await': 2,
+        'vars-on-top': 2,
+        'wrap-iife': 2
+        // 'yoda': 2
```
This condition name `yoda` still makes me laugh 😄

leventebalogh commented on best-practices.js:
```diff
@@ -0,0 +1,77 @@
+        // 'accessor-pairs': 2,
+        // 'array-callback-return': 2,
+        // 'block-scoped-var': 2,
+        // NO_RULE 'class-methods-use-this': 2,
```
I don't really agree that we should add all the rules we don't use commented out. This makes it harder to read, and requires you to read the README to understand why are they commented out and to get what `NO_RULE` means. Also we have to maintain the commented out rules as well as ESLint is updating.

I would rather just remove these rules for now (the ones we don't use) and add them back 1 by 1 by PRs when we feel the need for them. Then the conversation on that specific rule can be tracked down in the PR itself.

camilleriluke replied:
I the main reason I added this is for ease of use. 

Pre this PR someone would have to cross reference the rules with the eslint site and copy and paste the config files in here (check if already exists and so on). My intention was to make it easy for someone to look at the category file (f.ex `stylistic.js`) and get a fast clear picture of what are the possible rules they can enabled.

I will add the suggested workflow to follow in a contribution guide. That would give all the information to anyone who would like to contribute.

leventebalogh commented on index.js:
```diff
@@ -1,9 +1,18 @@
+        './ecma-script-6.js'
+    ],
+    rules: {},
     globals: {
```
I think globals setup should go with environments, e.g. we have different globals needed for a test environment, for a browser and a node environment.

leventebalogh reviewed (changes requested):
Our first PR 🚀  :1st_place_medal: 
Nice job! :)

leventebalogh commented:
I don't know the right solution yet, I am just starting out from that I would like to use something like this:

```javascript
// tests/.eslintrc
{
    "extends": [
        "casumo",
        "casumo/es6"
        "casumo/mocha"
    ]
}
```

```javascript
// gulp/.eslintrc
{
    "extends": [
        "casumo",
        "casumo/es6"
        "casumo/node"
    ]
}
```

So the environment specific stuff would be only in `casumo/node` and `casumo/mocha` in this case. What do you think about that?

Commit:
refactor(rules): Move rules from root to rules folder

Commit:
chore(.edtiroconfig): Add .editorconfig

Commit:
refactor(rules): Split rules into configuration for better composability

Commit:
refactor(rules): Match rule folder categories to eslint documentation

camilleriluke commented on best-practices.js:
```diff
@@ -0,0 +1,77 @@
+        // 'accessor-pairs': 2,
+        // 'array-callback-return': 2,
+        // 'block-scoped-var': 2,
+        // NO_RULE 'class-methods-use-this': 2,
```
Had a change of heart and completely changed the direction.

Now we have
- `./configurations` all configurations every application should ideally pick from here
- `./rules` rules grouped by plugin type 
- `./rules/eslint` eslint rules split by categories
- `./rules/mocha` relevant mocha rules

So you could use it like this:

By default it is es5 config

```js
{
    "extends": "casumo"
}
```

or

```js
{
    "extends": "casumo/configurations/es5-browser"
}
```

and in the gulp folder

```js
{
    "extends": "casumo/configurations/es5-node"
}
```

Also all rules are enabled (no commented rules), this is to prepare for the check for missing rules npm script

camilleriluke commented:
Had a change of heart and completely changed the direction.

Now we have
- `./configurations` all configurations every application should ideally pick from here
- `./rules` rules grouped by plugin type 
- `./rules/eslint` eslint rules split by categories
- `./rules/mocha` relevant mocha rules

So you could use it like this:

By default it is es5 config

```js
{
    "extends": "casumo"
}
```

or

```js
{
    "extends": "casumo/configurations/es5-browser"
}
```

and in the gulp folder

```js
{
    "extends": "casumo/configurations/es5-node"
}
```

Also all rules are enabled (no commented rules), this is to prepare for the check for missing rules npm script

Commit:
refactor(rules): Move deprecated rules inside the configuration extenders

Commit:
build(npm-scripts): Add `test` npm script

Commit:
chore(.npmignore): Add .npmignore

leventebalogh commented on configurations/es5-test.js:
```diff
@@ -0,0 +1,13 @@
+module.exports = {
+    extends: [
+        '../configurations/es5.js',
```
`./es5.js` would be enough?

camilleriluke replied:
haha I was following the pattern `../` to be consistent and easy on the eye 😄 

this looks ugly 😛 
```js
extends: [
    './es5.js',
    '../rules/mocha/on.js'
],
```

leventebalogh replied:
Ahh, ok, I thought it was accidental.
Yea, it's really easier to read that way and communicates the context. Agreed :)

leventebalogh commented on configurations/es5-node.js:
```diff
@@ -0,0 +1,12 @@
+module.exports = {
+    extends: [
+        '../configurations/es5.js',
```
`./es5.js` would be enough?

camilleriluke replied:
same as https://github.com/Casumo/eslint-config-casumo/pull/1#discussion_r93766130

leventebalogh commented on configurations/es5.js:
```diff
@@ -0,0 +1,25 @@
+        '../rules/eslint/strict-mode/on.js',
+        '../rules/eslint/stylistic-issues/on.js',
+        '../rules/eslint/variables/on.js',
+        '../rules/eslint/deprecated/off.js'
```
Why do we require in turned off rules here?
`'../rules/eslint/deprecated/off.js'`

camilleriluke replied:
To get a green light from `eslint-find-rules`. 

Currently it is throwing an error if the deprecated rules are not included.


Waiting on https://github.com/sarbbottam/eslint-find-rules/issues/172 to be resolved and then we can safely remove them

leventebalogh replied:
Ahh, ok, I see now. Can we add a comment or a todo to these, so they are describing why they are there?

camilleriluke replied:
Agree :)

leventebalogh commented on configurations/es6-node-test.js:
```diff
@@ -0,0 +1,18 @@
+module.exports = {
+    extends: [
+        '../configurations/es6-test.js'
```
`'./es6-test.js'`?

camilleriluke replied:
same as https://github.com/Casumo/eslint-config-casumo/pull/1#discussion_r93766130

leventebalogh commented on configurations/es6-node-test.js:
```diff
@@ -0,0 +1,18 @@
+    },
+    globals: {
+        expect: true,
+        sandbox: true
```
Shouldn't we add
```
    describe: true,
    it: true
```
as well?

camilleriluke replied:
`env mocha` does the trick :)

leventebalogh replied:
aaa, cool I missed that :P 👍

leventebalogh commented on configurations/es5.js:
```diff
@@ -0,0 +1,25 @@
+        '../rules/eslint/best-practices/on.js',
+        '../rules/eslint/possible-errors/on.js',
+        '../rules/eslint/ecma-script-6/off.js',
+        '../rules/eslint/node-js-and-common-js/off.js',
```
Is there any benefit of requiring in the turned off rules explicitly here?

camilleriluke replied:
Yes because the idea is to always ship a full set of rules including on and off. 

Like this we are explicitly ignoring `ecma-script-6` and `node-js-and-common-js` rules.

leventebalogh replied:
👍

leventebalogh commented on configurations/es6-node.js:
```diff
@@ -0,0 +1,24 @@
+module.exports = {
+    extends: [
+        '../configurations/es6.js',
```
Relative path `./es6.js`?

camilleriluke replied:
same as https://github.com/Casumo/eslint-config-casumo/pull/1#discussion_r93766130

leventebalogh commented on configurations/off.js:
```diff
@@ -0,0 +1,20 @@
+module.exports = {
```
Why do we need a composition file for the turned off rules?

camilleriluke replied:
This would be useful if someone wants to have only small number of rules turned on. Composing a config file with `off.js` and their overrides

leventebalogh replied:
Hmm... I think it is descriptive enough if you only import / define the rules that you would like to enable. As no eslint rules are enabled by default, when would someone need to import this?

camilleriluke replied:
> As no eslint rules are enabled by default, when would someone need to import this?

If you do not import all the off rules then `eslint-find-rules` would fail with the `-u` option

leventebalogh commented:
One thing that I miss is the contribution guideline for it, as I don't really get the main differences between `on` and `off` rules, and how to update / use them, so I think new people wouldn't get it as well. 

First I thought that the `off` rule files are containing rules that we don't use, however I saw that the `on` files contain also the same rules. Are they used to turn off categories explicitly?

Commit:
docs(README): Update docs

Commit:
refactor(rules): Add comment on why `deprecated/off` is needed

Commit:
refactor(rules): Update relative paths to `configurations` folder

Commit:
docs(README): Improve documentation

Updated contribution steps and added better explenation of the configuration structure

Commit:
chore(package.json): Bump version

Version bump to 1.1.0

camilleriluke commented:
PR changes complete. Bumped the version to `v1.1.0`.

Waiting for 👍 so I will merge it in and publish the package.
